xarray/core/dataset.py

```diff
@@ -2379,6 +2394,24 @@ def expand_dims(self, dim, axis=None):
             raise ValueError('dims should not contain duplicate values.')
 
         variables = OrderedDict()
+        if utils.is_dict_like(dim):
+            # If dim is a dict, then ensure that the values are either integers
+            # or iterables.
+            for k, v in dim.items():
+                if hasattr(v, "__iter__"):
+                    # If the value for the new dimension is an iterable, then
+                    # save the coordinates to the variables dict, and set the
+                    # value within the dim dict to the length of the iterable
+                    # for later use.
+                    variables[k] = xr.IndexVariable(k, list(v))
```

Comment:
I think `variables[k] = Variable((k, ), v)` is OK for here.

Reply:
We do currently try to use `IndexVariable` for every variable that gets an associated index. But I would definitely avoid casting to `list()`, since IndexVariable/Variable will internally convert it into a numpy.ndarray. So just `IndexVariable((k,), v)` would work.